File: app/src/main/java/com/nominacopro/domain/calculator/SettlementCalculator.kt
package com.nominacopro.domain.calculator

import com.nominacopro.domain.law.ColombiaLaborLaw2026
import com.nominacopro.domain.model.EmployeeProfile
import com.nominacopro.domain.model.LiquidationEstimate
import com.nominacopro.domain.model.SemesterSettlement
import com.nominacopro.domain.model.WorkDayEntry
import com.nominacopro.domain.model.YearSettlementReport
import java.time.LocalDate

/**
 * Prima de servicios, cesantías e intereses — CST arts. 249, 306–308; Ley 344/1996.
 * Base prestacional = salario + auxilio de transporte (si aplica).
 * Fórmula: base × días devengados ÷ 360 (año comercial laboral).
 */
object SettlementCalculator {

    fun calculate(
        profile: EmployeeProfile,
        year: Int,
        entries: List<WorkDayEntry>,
        manualHolidays: Set<LocalDate>,
        pendingVacationDays: Int = 0,
        asOf: LocalDate = LocalDate.now(),
    ): YearSettlementReport {
        val yearEnd = minOf(asOf, LocalDate.of(year, 12, 31))
        val yearStart = LocalDate.of(year, 1, 1)
        if (yearEnd.isBefore(yearStart)) {
            return emptyReport(year, pendingVacationDays, yearEnd)
        }

        val yearEntries = entries.filter { it.date.year == year && !it.date.isAfter(yearEnd) }

        val sem1Start = LocalDate.of(year, 1, 1)
        val sem1End = minOf(LocalDate.of(year, 6, 30), yearEnd)
        val sem2Start = LocalDate.of(year, 7, 1)
        val sem2End = yearEnd

        val firstSemester = semesterSettlement(
            profile, yearEntries, manualHolidays, sem1Start, sem1End,
            label = "Primer semestre (ene–jun)",
            deadline = "30 de junio",
        )
        val secondSemester = if (!yearEnd.isBefore(sem2Start)) {
            semesterSettlement(
                profile, yearEntries, manualHolidays, sem2Start, sem2End,
                label = "Segundo semestre (jul–dic)",
                deadline = "20 de diciembre",
            )
        } else {
            SemesterSettlement(
                label = "Segundo semestre (jul–dic)",
                start = sem2Start,
                end = LocalDate.of(year, 12, 31),
                workedDays = 0,
                remuneratedRestDays = 0,
                primaAmount = 0L,
                paymentDeadline = "20 de diciembre",
            )
        }

        val annualPayroll = PayrollEngine.liquidateDateRange(
            profile = profile,
            start = yearStart,
            end = yearEnd,
            entries = yearEntries,
            manualHolidays = manualHolidays,
            referenceYear = year,
            referenceMonth = 1,
        )
        val annualDays = annualPayroll.workedDays + annualPayroll.remuneratedRestDays
        val cesantias = prestacionesAmount(profile, annualDays)
        val intereses = (cesantias * ColombiaLaborLaw2026.INTERES_CESANTIAS_ANUAL * annualDays /
            ColombiaLaborLaw2026.DIAS_ANIO_PRESTACIONES).toLong()

        val currentSemesterPrima = if (!yearEnd.isBefore(sem2Start)) {
            secondSemester.primaAmount
        } else {
            firstSemester.primaAmount
        }

        val vacaciones = ColombiaLaborLaw2026.dailyRate(profile.monthlySalary)
            .toLong() * pendingVacationDays.coerceAtLeast(0)

        val liquidation = LiquidationEstimate(
            cesantias = cesantias,
            interesesCesantias = intereses,
            primaProporcional = currentSemesterPrima,
            vacaciones = vacaciones,
            pendingVacationDays = pendingVacationDays.coerceAtLeast(0),
            periodEnd = yearEnd,
            total = cesantias + intereses + currentSemesterPrima + vacaciones,
        )

        return YearSettlementReport(
            year = year,
            firstSemester = firstSemester,
            secondSemester = secondSemester,
            annualCesantias = cesantias,
            annualInteresesCesantias = intereses,
            liquidation = liquidation,
        )
    }

    private fun semesterSettlement(
        profile: EmployeeProfile,
        allEntries: List<WorkDayEntry>,
        manualHolidays: Set<LocalDate>,
        start: LocalDate,
        end: LocalDate,
        label: String,
        deadline: String,
    ): SemesterSettlement {
        if (end.isBefore(start)) {
            return SemesterSettlement(label, start, end, 0, 0, 0L, deadline)
        }
        val slice = allEntries.filter { it.date in start..end }
        val payroll = PayrollEngine.liquidateDateRange(
            profile = profile,
            start = start,
            end = end,
            entries = slice,
            manualHolidays = manualHolidays,
            referenceYear = start.year,
            referenceMonth = start.monthValue,
        )
        val days = payroll.workedDays + payroll.remuneratedRestDays
        return SemesterSettlement(
            label = label,
            start = start,
            end = end,
            workedDays = payroll.workedDays,
            remuneratedRestDays = payroll.remuneratedRestDays,
            primaAmount = prestacionesAmount(profile, days),
            paymentDeadline = deadline,
        )
    }

    private fun prestacionesAmount(profile: EmployeeProfile, days: Int): Long {
        if (days <= 0) return 0L
        val base = profile.monthlySalary +
            if (ColombiaLaborLaw2026.qualifiesTransport(profile.monthlySalary)) {
                ColombiaLaborLaw2026.SUBSIDIO_TRANSPORTE
            } else {
                0L
            }
        return (base * days / ColombiaLaborLaw2026.DIAS_ANIO_PRESTACIONES).toLong()
    }

    private fun emptyReport(year: Int, vacationDays: Int, end: LocalDate) = YearSettlementReport(
        year = year,
        firstSemester = SemesterSettlement("Primer semestre", LocalDate.of(year, 1, 1), LocalDate.of(year, 6, 30), 0, 0, 0L, "30 de junio"),
        secondSemester = SemesterSettlement("Segundo semestre", LocalDate.of(year, 7, 1), LocalDate.of(year, 12, 31), 0, 0, 0L, "20 de diciembre"),
        annualCesantias = 0L,
        annualInteresesCesantias = 0L,
        liquidation = LiquidationEstimate(0, 0, 0, 0, vacationDays, end, 0),
    )
}

File: web/lib/payroll/settlementCalculator.ts
import * as Law from "./colombiaLaborLaw2026";
import { isAfter, isBefore, type LocalDate } from "./dates";
import { liquidateDateRange } from "./payrollEngine";
import type {
  EmployeeProfile,
  LiquidationEstimate,
  ManualDeduction,
  SemesterSettlement,
  WorkDayEntry,
  YearSettlementReport,
} from "./models";

function prestacionesAmount(profile: EmployeeProfile, days: number): number {
  if (days <= 0) return 0;
  const base =
    profile.monthlySalary +
    (Law.qualifiesTransport(profile.monthlySalary) ? Law.SUBSIDIO_TRANSPORTE : 0);
  return Math.trunc((base * days) / Law.DIAS_ANIO_PRESTACIONES);
}

function emptyReport(
  year: number,
  vacationDays: number,
  end: LocalDate,
): YearSettlementReport {
  return {
    year,
    firstSemester: {
      label: "Primer semestre",
      start: `${year}-01-01`,
      end: `${year}-06-30`,
      workedDays: 0,
      remuneratedRestDays: 0,
      primaAmount: 0,
      paymentDeadline: "30 de junio",
    },
    secondSemester: {
      label: "Segundo semestre",
      start: `${year}-07-01`,
      end: `${year}-12-31`,
      workedDays: 0,
      remuneratedRestDays: 0,
      primaAmount: 0,
      paymentDeadline: "20 de diciembre",
    },
    annualCesantias: 0,
    annualInteresesCesantias: 0,
    liquidation: {
      cesantias: 0,
      interesesCesantias: 0,
      primaProporcional: 0,
      vacaciones: 0,
      pendingVacationDays: vacationDays,
      periodEnd: end,
      total: 0,
    },
  };
}

function semesterSettlement(
  profile: EmployeeProfile,
  allEntries: WorkDayEntry[],
  manualHolidays: Set<LocalDate>,
  start: LocalDate,
  end: LocalDate,
  label: string,
  deadline: string,
): SemesterSettlement {
  if (isAfter(start, end)) {
    return { label, start, end, workedDays: 0, remuneratedRestDays: 0, primaAmount: 0, paymentDeadline: deadline };
  }
  const slice = allEntries.filter((e) => !isAfter(e.date, end) && !isAfter(start, e.date));
  const payroll = liquidateDateRange(
    profile,
    start,
    end,
    slice,
    manualHolidays,
    parseInt(start.slice(0, 4), 10),
    parseInt(start.slice(5, 7), 10),
  );
  const days = payroll.workedDays + payroll.remuneratedRestDays;
  return {
    label,
    start,
    end,
    workedDays: payroll.workedDays,
    remuneratedRestDays: payroll.remuneratedRestDays,
    primaAmount: prestacionesAmount(profile, days),
    paymentDeadline: deadline,
  };
}

export function calculateSettlement(
  profile: EmployeeProfile,
  year: number,
  entries: WorkDayEntry[],
  manualHolidays: Set<LocalDate>,
  pendingVacationDays: number = 0,
  asOf: LocalDate = todayIso(),
): YearSettlementReport {
  const yearEnd = minDate(asOf, `${year}-12-31`);
  const yearStart = `${year}-01-01`;
  if (isAfter(yearStart, yearEnd)) {
    return emptyReport(year, pendingVacationDays, yearEnd);
  }

  const yearEntries = entries.filter((e) => e.date.startsWith(String(year)) && !isAfter(e.date, yearEnd));

  const sem1Start = `${year}-01-01`;
  const sem1End = minDate(`${year}-06-30`, yearEnd);
  const sem2Start = `${year}-07-01`;
  const sem2End = yearEnd;

  const firstSemester = semesterSettlement(
    profile,
    yearEntries,
    manualHolidays,
    sem1Start,
    sem1End,
    "Primer semestre (ene–jun)",
    "30 de junio",
  );

  const secondSemester = !isAfter(yearEnd, sem2Start)
    ? semesterSettlement(
        profile,
        yearEntries,
        manualHolidays,
        sem2Start,
        sem2End,
        "Segundo semestre (jul–dic)",
        "20 de diciembre",
      )
    : {
        label: "Segundo semestre (jul–dic)",
        start: sem2Start,
        end: `${year}-12-31`,
        workedDays: 0,
        remuneratedRestDays: 0,
        primaAmount: 0,
        paymentDeadline: "20 de diciembre",
      };

  const annualPayroll = liquidateDateRange(
    profile,
    yearStart,
    yearEnd,
    yearEntries,
    manualHolidays,
    year,
    1,
  );
  const annualDays = annualPayroll.workedDays + annualPayroll.remuneratedRestDays;
  const cesantias = prestacionesAmount(profile, annualDays);
  const intereses = Math.trunc(
    (cesantias * Law.INTERES_CESANTIAS_ANUAL * annualDays) / Law.DIAS_ANIO_PRESTACIONES,
  );

  const currentSemesterPrima = !isAfter(yearEnd, sem2Start)
    ? secondSemester.primaAmount
    : firstSemester.primaAmount;

  const vacaciones = Math.trunc(Law.dailyRate(profile.monthlySalary) * Math.max(0, pendingVacationDays));

  const liquidation: LiquidationEstimate = {
    cesantias,
    interesesCesantias: intereses,
    primaProporcional: currentSemesterPrima,
    vacaciones,
    pendingVacationDays: Math.max(0, pendingVacationDays),
    periodEnd: yearEnd,
    total: cesantias + intereses + currentSemesterPrima + vacaciones,
  };

  return {
    year,
    firstSemester,
    secondSemester,
    annualCesantias: cesantias,
    annualInteresesCesantias: intereses,
    liquidation,
  };
}

function todayIso(): LocalDate {
  const now = new Date();
  return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")}`;
}

function minDate(a: LocalDate, b: LocalDate): LocalDate {
  return a < b ? a : b;
}
